pliers/extractors/text.py

```diff
@@ -453,98 +446,282 @@ def __init__(self,
+            tokenizer, **self.tokenizer_kwargs)
+        super(BertExtractor, self).__init__()
+
+    def _mask(self, wds, mask):
+        return wds
+
+    def _preprocess(self, stims, mask):
+        els = [(e.text, e.onset, e.duration) for e in stims.elements]
+        wds, ons, dur = map(list, zip(*els))
+        tok = [self.tokenizer.tokenize(w) for w in self._mask(wds, mask)]
+        n_tok = [len(t) for t in tok]
+        wds, ons, dur = map(lambda x: np.repeat(x, n_tok), [wds, ons, dur])
+        tok = list(flatten(tok))
+        idx = self.tokenizer.encode(tok, return_tensors=self.framework)
+        return wds, ons, dur, tok, idx
 
     def _extract(self, stims):
-        if stims.name == '':
-            stims.name = ' '.join([s.text for s in stims.elements])
-
-        text, onsets, durations = zip(
-            *((s.text, s.onset, s.duration) for s in stims.elements))
-        tokens = [self.tokenizer.tokenize(t) for t in text]
-        tokens_flat = list(flatten(tokens))
-
-        def cast_to_token_level(word_level_list):
-            token_level_list = [itertools.repeat(word_level_list[i], len(tok))
-                                for i, tok in enumerate(tokens)]
-            token_level_list = list(flatten(token_level_list))
-            return token_level_list
-
-        t_text, t_ons, t_dur = map(cast_to_token_level,
-                                   [text, onsets, durations])
-        tensor_tokens = self.tokenizer.encode(tokens_flat,
-                                              return_tensors=self.framework)
-        output = self.model(tensor_tokens)
-        output = [out.detach().numpy() if self.framework == 'pt' else
-                  out.numpy() for out in output]
-
-        if self.encoding_level == 'token':
-            encoded_tokens = tokens_flat
-            encodings = output[0][:, 1:-1, :].squeeze()
-
-        elif self.encoding_level == 'sequence':
-            encoded_tokens = [' '.join(text)]
-            t_text = ['None']
-            if not any(val is None for val in [t_ons[-1],t_dur[-1],t_ons[0]]):
-                t_dur = t_ons[-1] + t_dur[-1] - t_ons[0]
-            else:
-                t_dur = None
-            t_ons = t_ons[0]
+        mask = self.mask or None
+        wds, ons, dur, tok, idx = self._preprocess(stims, mask=mask)
+        preds = self.model(idx)
+        preds = [p.detach() if self.framework == 'pt' else p for p in preds]
+        data, feat, ons, dur = self._postprocess(preds, tok, wds, ons, dur)
+        return ExtractorResult(data, stims, self, features=feat, onsets=ons, 
+                               durations=dur)
+
+    def _postprocess(self, preds, tok, wds, ons, dur):
+        out = preds[0][:, 1:-1, :].numpy().squeeze()
+        data = [out.tolist()]
+        feat = ['encoding']
+        if self.return_metadata:
+            data += [tok, wds]
+            feat += ['token', 'word']
+        return data, feat, ons, dur
+
+    def _get_model_attributes(self):
```

Comment:
This is a matter of preference, but since this always just returns a list, I would probably make it a class attribute called `_model_attributes` (instead of a class method) and just list it at the top of each class (much like `_log_attributes`).

Reply:
`_model_attributes` is now class attribute